src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/SdkDirectoryWorkloadManifestProvider.cs

```diff
@@ -264,7 +285,11 @@ void ThrowExceptionIfManifestsNotAvailable()
             }
         }
 
-        public IEnumerable<ReadableWorkloadManifest> GetManifests()
+        // The useInstallStateOnly flag indicates that, when performing garbage collection, we should not follow the normal logic of
+        // starting at manifest roots and following them to complete the closure. This means that the install state file has already
+        // been overwritten with exactly the set of manifests we want, and we want to keep only manifests in that list. Note that we
+        // keep the baseline manifests regardless.
+        public IEnumerable<ReadableWorkloadManifest> GetManifests(bool useInstallStateOnly = false)
```

Comment:
I'm not at all following what `useInstallStateOnly` is supposed to do.  The manifest roots used in the workload manifest provider are simply different directories where it should look for workload manifests.  This is used for scenarios like testing where you want to add a manifest, or with user local workload installs.  It's not at all related to garbage collection roots.

So, I think probably we should remove all of the `useInstallStateOnly` parameters.

Reply:
The idea is that I want it to _really_ go back to the particular version specified in the history record, which probably means garbage collecting manifests that might be earlier or later than that. This tries to throw out all our current heuristics and force it to exactly what was specified.

When I was first working on this, I thought we really had to get rid of later manifests, or else we'd failed. I now could be convinced if you want to argue that having manifests that weren't installed at the time the history record was written isn't a problem because they're side-by-side, and the resolver respects the install state file. In some sense, this felt like the more "burn everything wrong to the ground so only what we want remains" approach, which is perhaps a bit more than we want, but I do appreciate the simplicity of it.